cmd/applyconfig/applyconfig.go

```diff
@@ -84,24 +89,86 @@ func makeOcArgs(path, user string, dry bool) []string {
 		args = append(args, "--as", user)
 	}
 
-	return args
+	return exec.Command("oc", args...)
 }
 
-func apply(path, user string, dry bool) error {
-	args := makeOcArgs(path, user, dry)
+type executor interface {
+	runAndCheck(cmd *exec.Cmd, action string) ([]byte, error)
+}
+
+type commandExecutor struct{}
 
-	cmd := exec.Command("oc", args...)
-	if output, err := cmd.CombinedOutput(); err != nil {
-		if _, ok := err.(*exec.ExitError); ok {
-			fmt.Printf("ERROR: oc %s: failed to apply\n%s\n", strings.Join(args, " "), string(output))
+func (c commandExecutor) runAndCheck(cmd *exec.Cmd, action string) ([]byte, error) {
+	var output []byte
+	var err error
+	pretty := strings.Join(cmd.Args, " ")
+
+	if output, err = cmd.Output(); err != nil {
+		if exitError, ok := err.(*exec.ExitError); ok {
+			fmt.Printf("ERROR: %s: failed to %s\n%s\n", pretty, action, exitError.Stderr)
 		} else {
-			fmt.Printf("ERROR: oc %s: failed to execute: %v\n", strings.Join(args, " "), err)
+			fmt.Printf("ERROR: %s: failed to execute: %v\n", pretty, err)
 		}
-		return fmt.Errorf("failed to apply config")
+		return nil, fmt.Errorf("failed to %s config", action)
 	}
 
-	fmt.Printf("oc %s: OK\n", strings.Join(args, " "))
-	return nil
+	fmt.Printf("%s: OK\n", pretty)
+	return output, nil
+}
+
+type configApplier struct {
+	executor
+
+	path string
+	user string
+	dry  bool
+}
+
+func (c *configApplier) asGenericManifest() error {
+	cmd := makeOcApply(c.path, c.user, c.dry)
+	_, err := c.runAndCheck(cmd, "apply")
+	return err
+}
+
+func (c configApplier) asTemplate() error {
+	ocProcess := []string{"process", "-f", c.path}
+	if c.user != "" {
```

Comment:
Is it possible to factor this such that we only have one place that appends `--dry-run` and `--as` ?

Reply:
The difference here is that `oc process` does not have `--dry-run`, so the only common part is `--as`, but I can extract that, sure.

Reply:
Done